src/adapters/github.js

```diff
@@ -161,7 +162,57 @@ class GitHub extends PjaxAdapter {
 
     const repo = {username, reponame, branch, pullNumber};
     if (repo.branch) {
-      cb(null, repo);
+      if (isCodePage && type && type !== 'commit') {
+        // The window.location.pathname would be one of below forms:
+        //  - /username/repo/tree (or blob)/branch1
+        //  - /username/repo/tree (or blob)/branch1/folder1
+        //  - /username/repo/tree (or blob)/features/a (Gitflow workflow)
+        //  - /username/repo/tree (or blob)/features/a/folder1
+        // Reserve the branch name which is possibly mixing with the file path
+        const typeIdWithPath = match.slice(3);
+        let repoBranches, repoTags;
+
+        this._get('/branches', {repo, token}, (err, data) => {
+          repoBranches = err ? [] : data.map((item) => (item.name));
+
+          onResponseAvailable(err);
+        });
+
+        this._get('/tags', {repo, token}, (err, data) => {
+          repoTags = err ? [] : data.map((item) => (item.name));
+
+          onResponseAvailable(err);
+        });
+
+        function onResponseAvailable(err) {
+          if (!repoBranches || !repoTags) return;
```

Comment:
Should be done with util#parallel.

Reply:
Very concerned about these 2 calls. They happen in normal flow instead of exception flow. So network is hit every single time. Costly IMO. 

1. Isn't there any way to do what you're doing without hitting the network?
2. If has to hit network, must structure the code so that this is the last resort. For example, if `$('.branch-select-menu .select-menu-item.selected').data('name')` returns value, we have the branch name. Why have to go through all these extra network requests.